shell/browser/serial/serial_chooser_context.cc

```diff
@@ -0,0 +1,166 @@
+  auto it = ephemeral_ports_.find({requesting_origin, embedding_origin});
+  if (it != ephemeral_ports_.end()) {
+    const std::set<base::UnguessableToken> ports = it->second;
+    if (base::Contains(ports, port.token))
+      return true;
+  }
+
+  return false;
+}
+
+// static
+bool SerialChooserContext::CanStorePersistentEntry(
+    const device::mojom::SerialPortInfo& port) {
+  // If there is no display name then the path name will be used instead. The
+  // path name is not guaranteed to be stable. For example, on Linux the name
+  // "ttyUSB0" is reused for any USB serial device. A name like that would be
+  // confusing to show in settings when the device is disconnected.
+  if (!port.display_name || port.display_name->empty())
+    return false;
+
+  return port.persistent_id && !port.persistent_id->empty();
+}
+
+device::mojom::SerialPortManager* SerialChooserContext::GetPortManager() {
+  EnsurePortManagerConnection();
+  return port_manager_.get();
+}
+
+void SerialChooserContext::AddPortObserver(PortObserver* observer) {
+  port_observer_list_.AddObserver(observer);
+}
+
+void SerialChooserContext::RemovePortObserver(PortObserver* observer) {
+  port_observer_list_.RemoveObserver(observer);
+}
+
+base::WeakPtr<SerialChooserContext> SerialChooserContext::AsWeakPtr() {
+  return weak_factory_.GetWeakPtr();
+}
+
+void SerialChooserContext::OnPortAdded(device::mojom::SerialPortInfoPtr port) {
+  for (auto& observer : port_observer_list_)
+    observer.OnPortAdded(*port);
+}
+
+void SerialChooserContext::OnPortRemoved(
+    device::mojom::SerialPortInfoPtr port) {
+  for (auto& observer : port_observer_list_)
+    observer.OnPortRemoved(*port);
+
+  std::vector<std::pair<url::Origin, url::Origin>> revoked_url_pairs;
+  for (auto& map_entry : ephemeral_ports_) {
+    std::set<base::UnguessableToken>& ports = map_entry.second;
+    if (ports.erase(port->token) > 0) {
+      revoked_url_pairs.push_back(map_entry.first);
+    }
+  }
+
+  port_info_.erase(port->token);
+}
+
+void SerialChooserContext::EnsurePortManagerConnection() {
+  if (port_manager_)
+    return;
+
+  mojo::PendingRemote<device::mojom::SerialPortManager> manager;
+  content::GetDeviceService().BindSerialPortManager(
+      manager.InitWithNewPipeAndPassReceiver());
+  SetUpPortManagerConnection(std::move(manager));
+}
+
+void SerialChooserContext::SetUpPortManagerConnection(
+    mojo::PendingRemote<device::mojom::SerialPortManager> manager) {
+  port_manager_.Bind(std::move(manager));
+  port_manager_.set_disconnect_handler(
+      base::BindOnce(&SerialChooserContext::OnPortManagerConnectionError,
+                     base::Unretained(this)));
+
+  port_manager_->SetClient(client_receiver_.BindNewPipeAndPassRemote());
+}
+
+void SerialChooserContext::OnPortManagerConnectionError() {
+  port_manager_.reset();
+  client_receiver_.reset();
+
+  port_info_.clear();
+
+  std::vector<std::pair<url::Origin, url::Origin>> revoked_origins;
```

Comment:
We don't seem to do anything with this?

Reply:
yup... copied over from original implementation - I'll remove.